File: Helpers/src/main/java/com/dg/controls/SwipeRefreshLayoutEx.java
package com.dg.controls;

import android.content.Context;
import android.support.v4.widget.SwipeRefreshLayout;
import android.util.AttributeSet;

/**
 * Created by Daniel Cohen Gindi (danielgindi@gmail.com)
 */

public class SwipeRefreshLayoutEx extends SwipeRefreshLayout
{
    public SwipeRefreshLayoutEx(Context context)
    {
        super(context);
    }

    public SwipeRefreshLayoutEx(Context context, AttributeSet attrs)
    {
        super(context, attrs);
    }

    private boolean mMeasured = false;
    private boolean mHasPosponedSetRefreshing = false;
    private boolean mPreMeasureRefreshing = false;

    @Override
    public void onMeasure(int widthMeasureSpec, int heightMeasureSpec)
    {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);

        // There's a bug where setRefreshing(...) before onMeasure() has no effect at all.
        if (!mMeasured)
        {
            mMeasured = true;

            if (mHasPosponedSetRefreshing)
            {
                setRefreshing(mPreMeasureRefreshing);
            }
        }
    }


    @Override
    public void setRefreshing(boolean refreshing)
    {
        if (mMeasured)
        {
            mHasPosponedSetRefreshing = false;
            super.setRefreshing(refreshing);
        }
        else
        {
            // onMeasure was not called yet, so we postpone the setRefreshing until after onMeasure
            mHasPosponedSetRefreshing = true;
            mPreMeasureRefreshing = refreshing;
        }
    }
}

File: Helpers/src/main/java/com/dg/controls/SwipeRefreshLayoutEx.kt
package com.dg.controls

import android.content.Context
import android.support.v4.widget.SwipeRefreshLayout
import android.util.AttributeSet

class SwipeRefreshLayoutEx : SwipeRefreshLayout
{
    private var mMeasured = false
    private var mHasPostponedSetRefreshing = false
    private var mPreMeasureRefreshing = false

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int)
    {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)

        // There's a bug where setRefreshing(...) before onMeasure() has no effect at all.
        if (!mMeasured)
        {
            mMeasured = true

            if (mHasPostponedSetRefreshing)
            {
                isRefreshing = mPreMeasureRefreshing
            }
        }
    }

    override fun setRefreshing(refreshing: Boolean)
    {
        if (mMeasured)
        {
            mHasPostponedSetRefreshing = false
            super.setRefreshing(refreshing)
        }
        else
        {
            // onMeasure was not called yet, so we postpone the setRefreshing until after onMeasure
            mHasPostponedSetRefreshing = true
            mPreMeasureRefreshing = refreshing
        }
    }
}
